### ISSUE TITLE
[Bug] Infinite loop when running `dbt deps`

### ISSUE THREAD

**Issue by akerone** [dbt-labs/dbt-core#4377]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The command `dbt deps` can fall in an infinite loop when there are nested package dependencies.

I'm running Windows 10  with dbt installed in a venv.

Imagine 3 projects (p1, p2 and p3) in which p3 depends on p2, and p2 depends on p1 (p3 --> p2 --> p1).

All 3 projects exist in the same repository, different subdirectories. Using git to import each package.

Running `dbt deps` on p3 causes it to loop _ad infinitum_. 

Running `dbt deps` on p2 works as expected.

### Expected Behavior

The command `dbt deps` should resolve nested dependencies as there are no dependency loops.

## Option 1 - From scratch
1. In a blank folder run:
```
$ git init
$ mkdir dbt
$ cd dbt
$ dbt init p1
$ dbt init p2
$ dbt init p3
```
2. Go to each `dbt_project.yml` and change the name to p1, p2 or p3 accordingly
3. Add the following `packages.yml` to p2
```
packages:
  - git: 'https://github.com/<username>/<repo_name>.git'
    subdirectory: "dbt/p1"
```
4. Add the following `packages.yml` to p3
```
packages:
  - git: 'https://github.com/<username>/<repo_name>.git'
    subdirectory: "dbt/p2"
```
5. Run:
```
$ cd dbt/p3
$ dbt deps
```
6. ctrl-c when you're convinced it's infinite.

## Option 2 - Fest repo
1. Check out https://github.com/akerone/dbt-deps-poc/
2. Run the following commands
```
$ cd dbt/p3
$ dbt deps
```

### Relevant log output

```shell
\dbt\p3> dbt deps
Running with dbt=0.21.1
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!
...
...
...
WARNING: The git package "https://github.com/akerone/dbt-deps-poc.git" 
        is not pinned, using HEAD (default branch).
        This can introduce breaking changes into your project without warning!

See https://docs.getdbt.com/docs/package-management#section-specifying-package-versions
ctrl-c
```

### Environment

```markdown
- OS: Windows 10
- Python: 3.9.7 64-bit
- dbt: 0.21.1
```

### What database are you using dbt with?

bigquery

**Comment by akerone** [comment]
Update: Having the 3 dbt projects in different repos works fine.

**Comment by jtcohen6** [comment]
@akerone Thanks for the detailed write-up! I will say, this feels like a bit of an edge case.

> All 3 projects exist in the same repository, different subdirectories. Using git to import each package.

In the case where multiple projects must cohabitate in the same repository, is there a good reason not to import them as [`local` packages](https://docs.getdbt.com/docs/building-a-dbt-project/package-management#local-packages) (i.e. symlinks)? That feels like a much tighter approach, and one that doesn't risk the infinite loop you've outlined.

I'm inclined to close this as a `wontfix`, since there exist reasonable workarounds. If you disagree, and think this is an important capability that `dbt deps` ought to support, we can discuss further.

**Comment by akerone** [comment]
@jtcohen6 The reason for not importing them as local packages is that I work for an organization where most of us _don't have permanent admin rights to our Windows machines_. This means the same example projects will fail for me when running `dbt deps`

```
dbt\\p3> dbt deps
Running with dbt=0.21.0
Installing ../p2
Encountered an error:
[WinError 1314] A required privilege is not held by the client: 'REDACTED\\dbt-deps-poc\\dbt\\p2' -> 'dbt_modules\\p2'
```

The next best thing seemed to be to import them from git itself.

**Comment by jtcohen6** [comment]
@akerone Ah ok, fair point! I'm pretty lost when it comes to debugging symlinking on Windows (as evidenced by https://github.com/dbt-labs/dbt-core/issues/4372 too).

I'd be _very_ interested in a way to support `local` packages on Windows that doesn't require `admin` permissions, if you know of any potential approaches. In that sense, the infinite loop with multiple nested/self-referential `git` dependencies strikes me as an [XY problem](https://en.wikipedia.org/wiki/XY_problem).
